src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -0,0 +1,363 @@
+        of other estimates.
+
+        :return: vijs, viis all of which have a spurious J or not.
+        """
+
+        n_ims = viis.shape[0]
+        n_vijs = vijs.shape[0]
+        nchoose2 = int(n_ims * (n_ims - 1) / 2)
+        assert viis.shape[1:] == (3, 3), "viis must be 3x3 matrices."
+        assert vijs.shape[1:] == (3, 3), "vijs must be 3x3 matrices."
+        assert n_vijs == nchoose2, "There must be n_ims-choose-2 vijs."
+
+        # Determine relative handedness of vijs.
+        sign_ij_J = self._J_sync_power_method(vijs)
+        n_signs = len(sign_ij_J)
+        assert (
+            n_signs == n_vijs
+        ), f"There must be a sign associated with each vij. There are {n_signs} signs and {n_vijs} vijs."
+
+        # Synchronize vijs
+        J = np.diag((-1, -1, 1))
+        for i in range(n_signs):
+            if sign_ij_J[i] == -1:
+                vijs[i] = J @ vijs[i] @ J
+
+        # Synchronize viis
+        # We use the fact that if v_ii and v_ij are of the same handedness, then v_ii @ v_ij = v_ij.
+        # If they are opposite handed then Jv_iiJ @ v_ij = v_ij. We compare each v_ii against all
+        # previously synchronized v_ij to get a consensus on the handedness of v_ii.
+
+        # All pairs (i,j) where i<j
+        indices = np.arange(n_ims)
+        pairs = [(i, j) for idx, i in enumerate(indices) for j in indices[idx + 1 :]]
+
+        for i in range(n_ims):
+            vii = viis[i]
+            J_consensus = 0
+            for j in range(n_ims):
+                if j < i:
+                    idx = pairs.index((j, i))
+                    vji = vijs[idx]
+
+                    err1 = norm(vji @ vii - vji)
+                    err2 = norm(vji @ J @ vii @ J - vji)
+
+                elif j > i:
+                    idx = pairs.index((i, j))
+                    vij = vijs[idx]
+
+                    err1 = norm(vii @ vij - vij)
+                    err2 = norm(J @ vii @ J @ vij - vij)
+
+                else:
+                    continue
+
+                # Accumulate J consensus
+                if err1 < err2:
+                    J_consensus -= 1
+                else:
+                    J_consensus += 1
+
+            if J_consensus > 0:
+                viis[i] = J @ viis[i] @ J
+        return vijs, viis
+
+    def _estimate_third_rows(self, vijs, viis):
+        """
+        Find the third row of each rotation matrix given third row outer products.
+
+        :param vijs: An n-choose-2x3x3 array where each 3x3 slice holds the third rows
+        outer product of the corresponding pair of matrices.
+
+        :param viis: An nx3x3 array where the i-th 3x3 slice holds the outer product of
+        the third row of Ri with itself.
+
+        :param n_symm: The underlying molecular symmetry.
+
+        :return: vis, An n_imagesx3 matrix whose i-th row is the third row of the rotation matrix Ri.
+        """
+
+        n_ims = viis.shape[0]
```

Comment:
Is it possible to clean up where we're getting `n_ims` from in several of these functions? Since these are very complex and have lots of tricky variable names, it would be best to just get this information from the class whenever possible to reduce clutter

Reply:
Is there any situation where we'd be feeding these methods a subset of images? If so disregard my comments on this :)

Reply:
Yes. Using `self.n_img` everywhere now.